cudax/test/memory_resource/get_memory_resource.cu

```diff
@@ -0,0 +1,154 @@
+};
+TEST_CASE("Can call get_resource on a type with a get_resource method returns an rvalue", "[resource]")
+{
+  with_get_resource_rvalue val{};
+  auto&& res = ::cuda::experimental::get_memory_resource(val);
+  STATIC_REQUIRE(cuda::std::is_same_v<decltype(res), device_resource&&>);
+  CUDAX_CHECK(val.res_ == res);
+}
+
+struct with_get_resource_non_const
+{
+  device_resource res_{};
+
+  device_resource get_resource() noexcept
+  {
+    return res_;
+  }
+};
+TEST_CASE("Cannot call get_resource on a type with a non-const get_resource method", "[resource]")
+{
+  STATIC_REQUIRE(
+    !::cuda::std::is_invocable_v<::cuda::experimental::get_memory_resource_t, const with_get_resource_non_const&>);
+}
+
+struct env_with_query_const_ref
+{
+  device_resource res_{};
+
+  const device_resource& query(::cuda::experimental::get_memory_resource_t) const noexcept
+  {
+    return res_;
+  }
+};
+TEST_CASE("Can call get_resource on an env with a get_memory_resource query that returns a const lvalue", "[resource]")
+{
+  env_with_query_const_ref val{};
+  auto&& res = ::cuda::experimental::get_memory_resource(val);
+  STATIC_REQUIRE(cuda::std::is_same_v<decltype(res), const device_resource&>);
+  CUDAX_CHECK(val.res_ == res);
+}
+
+struct env_with_query_rvalue
+{
+  device_resource res_{};
+
+  device_resource query(::cuda::experimental::get_memory_resource_t) const noexcept
+  {
+    return res_;
+  }
+};
+TEST_CASE("Can call get_resource on an env with a get_memory_resource query that returns an rvalue", "[resource]")
+{
+  env_with_query_rvalue val{};
+  auto&& res = ::cuda::experimental::get_memory_resource(val);
+  STATIC_REQUIRE(cuda::std::is_same_v<decltype(res), device_resource&&>);
+  CUDAX_CHECK(val.res_ == res);
+}
+
+struct env_with_query_non_const
+{
+  device_resource res_{};
+
+  const device_resource& query(::cuda::experimental::get_memory_resource_t) noexcept
+  {
+    return res_;
+  }
+};
+TEST_CASE("Cannot call get_resource on an env with a non-const query", "[resource]")
+{
+  STATIC_REQUIRE(
+    !::cuda::std::is_invocable_v<::cuda::experimental::get_memory_resource_t, const env_with_query_non_const&>);
+}
+
+struct env_with_query_and_method
+{
+  device_resource res_{};
+
+  const device_resource& get_resource() const noexcept
+  {
+    return res_;
+  }
+
+  device_resource query(::cuda::experimental::get_memory_resource_t) const noexcept
+  {
+    return res_;
+  }
+};
+TEST_CASE("Can call get_resource on a type with both get_resource and query", "[resource]")
+{
+  env_with_query_and_method val{};
+  auto&& res = ::cuda::experimental::get_memory_resource(val);
+  STATIC_REQUIRE(cuda::std::is_same_v<decltype(res), const device_resource&>);
+  CUDAX_CHECK(val.res_ == res);
+}
+
+struct with_get_resource_non_async
+{
+  cudax::device_memory_resource res_{};
+
+  cudax::device_memory_resource get_resource() noexcept
+  {
+    return res_;
+  }
+};
+TEST_CASE("Cannot call get_resource on an env with a non-async resource", "[resource]")
```

Comment:
I think this case is not valid anymore, since all resources we provide are asynchronous

Reply:
Oh yeah, this is failing because get_resource is not const we need a special non async resource